OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,304 @@
+		}
+
+		void ClientConnected(object sender, EventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				AddChatLine("Connected.");
+
+				var client = (IrcClient)sender;
+				client.Channels.Join("#" + Game.Settings.Irc.Channel);
+				AddChatLine("Joining channel #{0}...".F(Game.Settings.Irc.Channel));
+			});
+		}
+
+		void ClientDisconnected(object sender, EventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				AddChatLine("Disconnected.");
+
+				if (client != null)
+				{
+					client.Dispose();
+					client = null;
+				}
+
+				MaybeShowConnectPanel();
+			});
+		}
+
+		void ClientRegistered(object sender, EventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				var client = (IrcClient)sender;
+
+				client.LocalUser.MessageReceived += MessageReceived;
+				client.LocalUser.JoinedChannel += JoinedChannel;
+				client.LocalUser.LeftChannel += LeftChannel;
+
+				AddChatLine("Registered client.");
+			});
+		}
+
+		void NoticeReceived(object sender, IrcMessageEventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				AddChatLine("[{0}] NOTICE -{1}- {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), e.Source, e.Text));
+			});
+		}
+
+		void MessageReceived(object sender, IrcMessageEventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				AddChatLine("[{0}] <{1}> {2}".F(DateTime.Now.ToString(Game.Settings.Irc.TimestampFormat), e.Source.Name, e.Text));
+			});
+		}
+
+		void JoinedChannel(object sender, IrcChannelEventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				AddChatLine("Channel joined.");
+
+				e.Channel.TopicChanged += TopicChanged;
+				e.Channel.UserJoined += UserJoined;
+				e.Channel.UserLeft += UserLeft;
+				e.Channel.MessageReceived += MessageReceived;
+				e.Channel.UsersListReceived += UsersListReceived;
+			});
+		}
+
+		void TopicChanged(object sender, IrcUserEventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				AddChatLine("*** Topic: {0}".F(client.Channels.First().Topic));
+			});
+		}
+
+		void UsersListReceived(object sender, EventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				foreach (var client in client.Channels.First().Users)
+					nicknamePanel.AddChild(MakeLabelWidget(nicknameTemplate, client.User.NickName));
+			});
+		}
+
+		void LeftChannel(object sender, IrcChannelEventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				AddChatLine("Channel left");
+
+				e.Channel.UserJoined -= UserJoined;
+				e.Channel.UserLeft -= UserLeft;
+				e.Channel.MessageReceived -= MessageReceived;
+			});
+		}
+
+		void UserLeft(object sender, IrcChannelUserEventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				var name = ((IrcChannel)sender).Name;
+				AddChatLine("{0} left {1}".F(e.ChannelUser, name));
+				nicknamePanel.RemoveChild(nicknamePanel.Children.OfType<LabelWidget>().First(f => f.Text == name));
+			});
+		}
+
+		void UserJoined(object sender, IrcChannelUserEventArgs e)
+		{
+			Game.RunAfterTick(() => {
+				var nickName = e.ChannelUser.User.NickName;
+				AddChatLine("{0} joined in {1}".F(nickName, ((IrcChannel)sender).Name));
```

Comment:
I don't think we need to show channel names in the UI. There's only one channel supported by this code and the abstraction of IRC towards the generic 'global chat' where everyone is in one channel doesn't really warrant it.